### PR TITLE
Async compatible `StaticFilesPanel`

### PR DESCRIPTION
**Pull request by salty-ivy**
# Description

Uses signals for automatic instrumentation of `StaticFilesPanel` and trigger recording mechanis.
and 2 `ContextVar` one for storing staticfiles particular to that request context and another for to identify signal for each request context.

# Checklist:

- [x] I have added the relevant tests for this change.
- [x] I have added an item to the Pending section of ``docs/changes.rst``.

### CODE REVIEW

File: debug_toolbar/panels/staticfiles.py [hunk 1]
```diff
@@ -29,7 +31,9 @@ def url(self):
 
 
 # This will collect the StaticFile instances across threads.
-used_static_files = ContextVar("djdt_static_used_static_files")
+used_static_files = ContextVar("djdt_static_used_static_files", default=[])
```
Comment (tim-schilling) [inline debug_toolbar/panels/staticfiles.py:34]:
```suggestion
used_static_files = ContextVar("djdt_static_used_static_files", default=list)
```
Do we run the same risk here as we would with defining a default on a method argument of `[]` versus `None` and instantiating the list?

File: debug_toolbar/panels/staticfiles.py [hunk 2]
```diff
@@ -88,12 +97,25 @@ def __init__(self, *args, **kwargs):
...
+            # as sender of the signal and this handler will have multiple
+            # concurrent connections and we want to avoid storing of same
+            # staticfile from other connections as well.
+            if request_id_context_var.get() == self.request_id:
+                used_static_files.get().append(staticfile)
 
     def enable_instrumentation(self):
         storage.staticfiles_storage = DebugConfiguredStorage()
+        record_static_file_signal.connect(self._store_static_files_signal_handler)
+        request_id_context_var.set(self.request_id)
```
Comment (tim-schilling) [inline debug_toolbar/panels/staticfiles.py:115]: We'll want to capture the [returned token](https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.set) here and then use it to [reset](https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.reset) the `ContextVar`

**Review by tim-schilling — APPROVED** [review]
I like this a lot!

**Comment by salty-ivy** [comment]
I think it still need those integration type tests.

File: debug_toolbar/panels/staticfiles.py [hunk 3]
```diff
@@ -87,12 +97,28 @@ def __init__(self, *args, **kwargs):
...
+        # that was used to create the panel.
+        # as sender of the signal and this handler will have multiple
+        # concurrent connections and we want to avoid storing of same
+        # staticfile from other connections as well.
+        if request_id_context_var.get() == self.request_id:
+            self.used_paths.append(staticfile)
+
+    def enable_instrumentation(self):
+        record_static_file_signal.connect(self._store_static_files_signal_handler)
+        self.ctx_token = request_id_context_var.set(self.request_id)
```
Comment (tim-schilling) [inline debug_toolbar/panels/staticfiles.py:117]:
```suggestion
        self.ctx_token = request_id_context_var.set(self.request_id)
        record_static_file_signal.connect(self._store_static_files_signal_handler)
```

File: tests/panels/test_staticfiles.py [hunk 4]
```diff
@@ -27,6 +29,17 @@ def test_default_case(self):
...
+    async def test_store_staticfiles_with_async_context(self):
+        async def get_response(request):
+            # template contains one static file
+            return render(request, "staticfiles/async_static.html")
+
+        self._get_response = get_response
+        async_request = AsyncRequestFactory().get("/")
+        response = await self.panel.process_request(async_request)
+        self.panel.generate_stats(self.request, response)
+        self.assertNotEqual(self.panel.num_used, 0)
```
Comment (tim-schilling) [inline tests/panels/test_staticfiles.py:41]:
```suggestion
        self.assertEqual(self.panel.num_used, 1)
```

**Review by tim-schilling — APPROVED** [review]
Looks good. I have minor changes I'd like to see.

**Pull request merged**